L1Trigger/DTTriggerPhase2/src/MuonPathAssociator.cc

```diff
@@ -0,0 +1,1102 @@
+  useBX_correlation = pset.getUntrackedParameter<bool>("useBX_correlation");
+  allow_confirmation = pset.getUntrackedParameter<bool>("allow_confirmation");
+  dT0_correlate_TP = pset.getUntrackedParameter<double>("dT0_correlate_TP");
+  dBX_correlate_TP = pset.getUntrackedParameter<int>("dBX_correlate_TP");
+  dTanPsi_correlate_TP = pset.getUntrackedParameter<double>("dTanPsi_correlate_TP");
+  minx_match_2digis = pset.getUntrackedParameter<double>("minx_match_2digis");
+  chi2corTh = pset.getUntrackedParameter<double>("chi2corTh");
+
+  if (debug)
+    cout << "MuonPathAssociator: constructor" << endl;
+
+  //shift
+  int rawId;
+  shift_filename = pset.getParameter<edm::FileInPath>("shift_filename");
+  std::ifstream ifin3(shift_filename.fullPath());
+  double shift;
+  if (ifin3.fail()) {
+    throw cms::Exception("Missing Input File")
+        << "MuonPathAnalyzerPerSL::MuonPathAnalyzerPerSL() -  Cannot find " << shift_filename.fullPath();
+  }
+  while (ifin3.good()) {
+    ifin3 >> rawId >> shift;
+    shiftinfo[rawId] = shift;
+  }
+
+  dtGeomH = iC.esConsumes<DTGeometry, MuonGeometryRecord, edm::Transition::BeginRun>();
+}
+
+MuonPathAssociator::~MuonPathAssociator() {
+  if (debug)
+    cout << "MuonPathAssociator: destructor" << endl;
+}
+
+// ============================================================================
+// Main methods (initialise, run, finish)
+// ============================================================================
+void MuonPathAssociator::initialise(const edm::EventSetup &iEventSetup) {
+  if (debug)
+    cout << "MuonPathAssociator::initialiase" << endl;
+
+  const MuonGeometryRecord &geom = iEventSetup.get<MuonGeometryRecord>();
+  dtGeo_ = &geom.get(dtGeomH);
+}
+
+void MuonPathAssociator::run(edm::Event &iEvent,
+                             const edm::EventSetup &iEventSetup,
+                             edm::Handle<DTDigiCollection> digis,
+                             std::vector<metaPrimitive> &inMPaths,
+                             std::vector<metaPrimitive> &outMPaths) {
+  if (dT0_correlate_TP)
+    correlateMPaths(digis, inMPaths, outMPaths);
+  else {
+    for (auto metaPrimitiveIt = inMPaths.begin(); metaPrimitiveIt != inMPaths.end(); ++metaPrimitiveIt)
+      outMPaths.push_back(*metaPrimitiveIt);
+  }
+}
+
+void MuonPathAssociator::finish() {
+  if (debug)
+    cout << "MuonPathAssociator: finish" << endl;
+};
+
+void MuonPathAssociator::correlateMPaths(edm::Handle<DTDigiCollection> dtdigis,
+                                         std::vector<metaPrimitive> &inMPaths,
+                                         std::vector<metaPrimitive> &outMPaths) {
+  int x_prec_inv = (int)(1. / (10. * x_precision));
+  int numberOfBits = (int)(round(std::log(x_prec_inv) / std::log(2.)));
+
+  //Silvia's code for correlationg filteredMetaPrimitives
+
+  if (debug)
+    std::cout << "starting correlation" << std::endl;
+
+  for (int wh = -2; wh <= 2; wh++) {
```

Comment:
magic numbers

Reply:
these are wheel, sector and station numbers, added a comment to make the code more readable.